websockets/connection.go

```diff
@@ -0,0 +1,181 @@
+	if err != nil {
+		return nil, err
+	}
+
+	conn, _, err := websocket.DefaultDialer.Dial(u.String(), http.Header{})
+	if err != nil {
+		return nil, err
+	}
+	return conn, nil
+}
+
+func newConnection(
+	logger polylog.Logger,
+	conn *websocket.Conn,
+	source messageSource,
+	msgChan chan message,
+	stopChan chan error,
+) *connection {
+	c := &connection{
+		logger:   logger,
+		Conn:     conn,
+		source:   source,
+		msgChan:  msgChan,
+		stopChan: stopChan,
+	}
+
+	go c.connLoop()
+	go c.pingLoop()
+
+	return c
+}
+
+// connLoop reads messages from the websocket connection and sends them to the bridge's msgChan
+func (c *connection) connLoop() {
+	for {
+		select {
+		case err := <-c.stopChan:
+			if err := c.cleanup(err); err != nil {
+				c.logger.Error().Err(err).Msg("error cleaning up connection")
+			}
+			return
+
+		default:
+			messageType, msg, err := c.ReadMessage()
+			if err != nil {
+				c.handleError(err, c.source)
+				return
+			}
+
+			c.msgChan <- message{
+				data:        msg,
+				source:      c.source,
+				messageType: messageType,
+			}
+		}
+	}
+}
+
+// cleanup closes the client and gateway connections
+func (c *connection) cleanup(err error) error {
+	closeMsg := websocket.FormatCloseMessage(websocket.CloseNormalClosure, err.Error())
+
+	// Close the connection and send a reason for the closure
+	if err := c.WriteMessage(websocket.CloseMessage, closeMsg); err != nil {
+		c.logger.Error().Err(err).Msg("error writing close message to connection")
+		return err
+	}
+	if err := c.Close(); err != nil {
+		c.logger.Error().Err(err).Msg("error closing connection")
+		return err
+	}
+
+	return nil
+}
+
+// handleError handles errors from the websocket connection and sends them to the stopChan if applicable
+func (c *connection) handleError(err error, source messageSource) {
+	if websocket.IsCloseError(err, websocket.CloseNoStatusReceived) {
+		c.logger.Info().Msgf("%s connection closed by peer", source)
+	} else {
+		c.logger.Error().Err(err).Msgf(" %s error reading from connection", source)
+	}
+
+	select {
+	case <-c.stopChan:
+		// stopChan is already closed, do nothing
+	default:
+		// stopChan is still open, send the error
+		c.stopChan <- fmt.Errorf("error reading from %s connection: %w", source, err)
+	}
+}
+
+// pingLoop sends keep-alive ping messages to the connection and handles pong messages
+func (c *connection) pingLoop() {
+	ticker := time.NewTicker(pingPeriod)
+	defer func() {
+		ticker.Stop()
+	}()
+
+	initPingLoop := func() {
+		// Set initial read deadline
+		if err := c.SetReadDeadline(time.Now().Add(pongWait)); err != nil {
+			c.logger.Error().Err(err).Msg("failed to set initial read deadline")
+		}
+		// Extend read deadline on pong response
+		c.SetPongHandler(func(string) error {
+			if err := c.SetReadDeadline(time.Now().Add(pongWait)); err != nil {
+				c.logger.Error().Err(err).Msg("failed to set pong handler read deadline")
+			}
+
+			return nil
+		})
+	}
+
+	initPingLoop()
```

Comment:
I don't fully understand why this is a separate function?  It's not even called in a go-routine.

Reply:
That's a holdover from the previous implementation where it was called in multiple places (ie. on reconnect) but I've refactored the need for that out since we are offloading that responsibility to the client. So you are right it does not need to be a separate func; I have updated to remove the func var.